pkg/ingestor/parser/vuln/vuln.go

```diff
@@ -0,0 +1,179 @@
+}
+
+// NewVulnCertificationParser initializes the parser
+func NewVulnCertificationParser() common.DocumentParser {
+	return &parser{}
+}
+
+// Parse breaks out the document into the graph components
+func (c *parser) Parse(ctx context.Context, doc *processor.Document) error {
+	statement, err := parseVulnCertifyPredicate(doc.Blob)
+	if err != nil {
+		return fmt.Errorf("failed to parse slsa predicate: %w", err)
+	}
+	ps, ios, err := parseSubject(statement)
+	if err != nil {
+		return fmt.Errorf("unable to parse subject of statement: %w", err)
+	}
+	c.packages = ps
+	c.isOccs = ios
+	c.vulnData = parseMetadata(statement)
+	vs, ivs, err := parseVulns(statement)
+	if err != nil {
+		return fmt.Errorf("unable to parse vulns of statement: %w", err)
+	}
+	c.vulns = vs
+	c.isVulns = ivs
+	return nil
+}
+
+func parseVulnCertifyPredicate(p []byte) (*attestation_vuln.VulnerabilityStatement,
+	error) {
+	predicate := attestation_vuln.VulnerabilityStatement{}
+	if err := json.Unmarshal(p, &predicate); err != nil {
+		return nil, err
+	}
+	return &predicate, nil
+}
+
+func parseSubject(s *attestation_vuln.VulnerabilityStatement) ([]*generated.PkgInputSpec,
+	[]assembler.IsOccurenceIngest, error) {
+	var ps []*generated.PkgInputSpec
+	var ios []assembler.IsOccurenceIngest
+	for _, sub := range s.StatementHeader.Subject {
+		p, err := helpers.PurlToPkg(sub.Name)
+		if err != nil {
+			return nil, nil, fmt.Errorf("bad purl in statement header: %w", err)
+		}
+		ps = append(ps, p)
+		for a, d := range sub.Digest {
+			io := assembler.IsOccurenceIngest{
+				Pkg: p,
+				Artifact: &generated.ArtifactInputSpec{
+					Algorithm: a,
+					Digest:    d,
+				},
+				IsOccurence: &generated.IsOccurrenceInputSpec{
+					Justification: "Reported in vulnerability certifier", // FIXME is this ok?
+				},
+			}
+			ios = append(ios, io)
+		}
+	}
+	return ps, ios, nil
+}
+
+func parseMetadata(s *attestation_vuln.VulnerabilityStatement) *generated.VulnerabilityMetaDataInput {
+	return &generated.VulnerabilityMetaDataInput{
+		TimeScanned:    *s.Predicate.Metadata.ScannedOn,
+		DbUri:          s.Predicate.Scanner.Database.Uri,
+		DbVersion:      s.Predicate.Scanner.Database.Version,
+		ScannerUri:     s.Predicate.Scanner.Uri,
+		ScannerVersion: s.Predicate.Scanner.Version,
+	}
+}
+
+func parseVulns(s *attestation_vuln.VulnerabilityStatement) ([]*generated.OSVInputSpec,
+	[]assembler.IsVulnIngest, error) {
+	var vs []*generated.OSVInputSpec
+	var ivs []assembler.IsVulnIngest
+	for _, id := range s.Predicate.Scanner.Result {
+		v := &generated.OSVInputSpec{
+			OsvId: id.VulnerabilityId,
+		}
+		vs = append(vs, v)
+		cve, ghsa, err := helpers.OSVToGHSACVE(id.VulnerabilityId)
+		if err != nil {
+			return nil, nil, fmt.Errorf("could not parse vuln id in attestation: %w", err)
+		}
+		iv := assembler.IsVulnIngest{
+			OSV:  v,
+			CVE:  cve,
+			GHSA: ghsa,
+			IsVuln: &generated.IsVulnerabilityInputSpec{
+				Justification: "Decoded OSV ID", // FIXME is this ok?
```

Comment:
And here.

Reply:
Yea that's fine, maybe "data" instead of "ID"